lib/common/recurrence.dart

```diff
@@ -1,35 +1,98 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
+
 /// Class to bundle data for a task/event's recurrance
 class Recurrence {
-  bool enabled;
-  DateTime? timeStart;
-  DateTime? timeEnd;
+  late bool enabled;
+  late DateTime timeStart;
+  late DateTime timeEnd;
   // monday, tuesday, ... , saturday, sunday
-  List<bool>? dates;
-  int? id; // ideally this should be final
+  late List<bool> dates;
+  late String id; // ideally this should be final
 
   Recurrence(
+      {this.enabled = false,
+        DateTime? timeStart,
+        DateTime? timeEnd,
+        List<bool>? dates,
+        String? id}) {
+    this.timeStart = timeStart ?? DateTime(2000);
+    this.timeEnd = timeEnd ?? DateTime(2000);
+    this.dates = dates ?? [false, false, false, false, false, false, false];
+    this.id = id ?? DateTime.now().millisecondsSinceEpoch.toString();
+  }
+  
+  /// Set nullOrId: null to set id to auto-generate an id instead
+  Recurrence.requireFields(
       {required this.enabled,
         required this.timeStart,
         required this.timeEnd,
         required this.dates,
-        this.id}) {
-    id = id ?? DateTime.now().millisecondsSinceEpoch;
+        required String? nullOrId}) {
+    this.id = nullOrId ?? DateTime.now().millisecondsSinceEpoch.toString();
   }
 
-  Recurrence.requireFields(
-      {required this.enabled,
-        required this.timeStart,
-        required this.timeEnd,
-        required this.dates}) {
-    id ??= DateTime.now().millisecondsSinceEpoch; // something about this design pattern looks off... not sure how to make this better since flutter won't allow me to put an optional named argument in
+  Recurrence.fromMap(Map<String, dynamic>? recurrenceRulesMap, {String? extraId}) {
+    if (recurrenceRulesMap == null) {
+      enabled = false;
+      timeStart = DateTime(2000);
```

Comment:
Is there a way to set the default timeStart to something that doesn't mean a time? I feel like these dates may be misleading/hard to deal with. Someone/somewhere will have to detect if timeStart and timeEnd == DateTime(2000), which sounds ugly.

Reply:
although this might not really be a real concern as long as it's caught and the date is something like the epoch of time

Reply:
I tried to set it to year 0 but this was negative time. What about make a method like:
`setTimes(t1, t2)` that will enable set timeStart and timeEnd, then set `enabled=true` in there?

Reply:
@stxue1 Made a new commit, how does it look now? Also added new methods.

Reply:
I think it works, I doubt anyone will even try to scroll to 1970